### ISSUE TITLE
AWS s3 protocol slower

### ISSUE THREAD

**Issue by jbe456** [atoti/atoti#26]
## Description

I tried to load remote data using several methods and using the s3 protocol appears 200 times slower than using wget or pandas:
|method|timing|
|---|---|
|pandas http + read_pandas|96.8ms|
|wget http + read_csv|106ms|
|read_csv + s3|20.2s|

![image](https://user-images.githubusercontent.com/4415696/80513048-c4de2980-897e-11ea-8b9b-09fa8835294c.png)

## Reproduce

Install `python-wget` and run the attached notebook [main.zip](https://github.com/atoti/atoti/files/4547233/main.zip)

## Context

- Atoti: 0.4.0.20200428135859
- Python: 3.7.6
- Operating System: linux

## Logs (if relevant)

Here are the [logs](https://github.com/atoti/atoti/files/4547287/logs.txt)

**Comment by jbe456** [comment]
When timing the cell once, all the time seems to be spent downloading the data: `Wall time: 25.1s`

![image](https://user-images.githubusercontent.com/4415696/80515537-687d0900-8982-11ea-9bb1-ef09bfeee4a3.png)


The bucket is located in `EU (Paris) eu-west-3` and my connection however is good:
![image](https://user-images.githubusercontent.com/4415696/80515810-d6293500-8982-11ea-8a30-a52ed515cc9d.png)

The ping also seems good:
```console
ζ ping data.atoti.io -c 5                                                                                                              
PING d3e1owrtdx1nmn.cloudfront.net (13.225.31.56) 56(84) bytes of data.
64 bytes from server-13-225-31-56.cdg3.r.cloudfront.net (13.225.31.56): icmp_seq=1 ttl=242 time=17.9 ms
64 bytes from server-13-225-31-56.cdg3.r.cloudfront.net (13.225.31.56): icmp_seq=2 ttl=242 time=6.09 ms
64 bytes from server-13-225-31-56.cdg3.r.cloudfront.net (13.225.31.56): icmp_seq=3 ttl=242 time=5.86 ms
64 bytes from server-13-225-31-56.cdg3.r.cloudfront.net (13.225.31.56): icmp_seq=4 ttl=242 time=6.09 ms
64 bytes from server-13-225-31-56.cdg3.r.cloudfront.net (13.225.31.56): icmp_seq=5 ttl=242 time=6.12 ms

--- d3e1owrtdx1nmn.cloudfront.net ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4005ms
rtt min/avg/max/mdev = 5.860/8.409/17.890/4.741 ms
```
```console
ζ ping data.atoti.io.s3-website.eu-west-3.amazonaws.com -c 5                                                                           
PING s3-website.eu-west-3.amazonaws.com (52.95.154.22) 56(84) bytes of data.
64 bytes from s3-website.eu-west-3.amazonaws.com (52.95.154.22): icmp_seq=1 ttl=48 time=5.35 ms
64 bytes from s3-website.eu-west-3.amazonaws.com (52.95.154.22): icmp_seq=2 ttl=48 time=5.31 ms
64 bytes from s3-website.eu-west-3.amazonaws.com (52.95.154.22): icmp_seq=3 ttl=48 time=6.41 ms
64 bytes from s3-website.eu-west-3.amazonaws.com (52.95.154.22): icmp_seq=4 ttl=48 time=9.82 ms
64 bytes from s3-website.eu-west-3.amazonaws.com (52.95.154.22): icmp_seq=5 ttl=48 time=5.74 ms

--- s3-website.eu-west-3.amazonaws.com ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4007ms
rtt min/avg/max/mdev = 5.313/6.526/9.821/1.693 ms

```

**Comment by jbe456** [comment]
Using `pandas` and `s3fs` I have 62.5ms:

![image](https://user-images.githubusercontent.com/4415696/80518138-5ef5a000-8986-11ea-92d5-79664c87848c.png)

**Comment by HuifangYeo** [comment]
From sg:
![image](https://user-images.githubusercontent.com/35210495/80550917-46bc6a00-89f4-11ea-910d-fde574c590ae.png)

But the direct read_csv using s3 addressing scheme is still very slow compared to the other 2:
![image](https://user-images.githubusercontent.com/35210495/80550710-86cf1d00-89f3-11ea-83bd-b45d9dfda9fb.png)

**Comment by SamTaz** [comment]
Pretty strange I do not have the same results, the s3 is a bit slower but only 2 times:

![image](https://user-images.githubusercontent.com/39309622/80573206-dadffe80-89ff-11ea-9000-1044becd96fd.png)

Tested from Paris in WiFi:
![image](https://user-images.githubusercontent.com/39309622/80573335-0bc03380-8a00-11ea-9424-ad7e9cbd907c.png)

**Comment by fabiencelier** [comment]
Hello, we have identified the issue and we will fix it

Here are some technical details to understand why only some people are concerned:
There is an issue with the credentials providers of our S3 client. If you are logged in AWS (either with a .aws/credential file or with the AWS environment variables crendentials) the client uses these credentials and it's fast.
If you are not logged in then it fallbacks on another slow credentials provider that checks if you are on an EC2 instance and timeouts after ~2s. This happens at each request to S3 creating the ~20s delay you see. 
If you want to have fast S3 before the fix I think you can simply log in to any AWS account.

**Comment by fabiencelier** [comment]
We don't want to get rid of this credentials provider as it is very convenient when running on AWS EC2.

Pandas is relying on [s3fs](https://github.com/dask/s3fs/) which is based on botocore. They have the same way of [fallbacking on credential providers](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#configuring-credentials) which ends with the "Instance metadata service on an Amazon EC2"

If you install s3fs and do this without any credentials:
```python
import pandas as pd
```
```python
%%time
df = pd.read_csv("s3://data.atoti.io/notebooks/collateral-shortfall-monitoring/assets_positions.csv")
df.head()
```
You will see that the first call lasts for 2sec while any following call is ~100ms. This means they also have this ~2s timeout but then cache the result.

We should do something similar and only try to retrieve these credentials once

**Comment by HuifangYeo** [comment]
Hi,
it seems like not just the initial read is slow but joining the store and creating simulation are slower in the version using s3. For instance, if you look at the gif below, the left side uses s3 source while the right side downloads file using wget before processing.

![slowness_s3](https://user-images.githubusercontent.com/35210495/81146731-70a0fe00-8fab-11ea-8b70-81a3d81f049e.gif)

**Comment by fabiencelier** [comment]
The loading of the data can happen at several places (read_csv, join, create_cube...). For instance when creating a join between 2 stores we need to rebuild the datastore and re-trigger the loading of the data from S3.

With the fix available in the next release you will only have a single 2s delay per session, the first time you query S3 (probably at the read_csv).

The fix is already available in the continuous releases (since 2020-05-02 I think) if you want to try it and confirm that it fixes your issue

**Comment by HuifangYeo** [comment]
Hi @fabiencelier,
I thought during read_csv, the data is already loaded into memory and used for the future processing such as join, create_cube or simulations. Why is there a need to re-trigger the loading of data?

**Comment by fabiencelier** [comment]
This an internal limitation, the datastore cannot be modified so you need to recreate and reload it every time the structure changes.
This will change in the future.

**Comment by jbe456** [comment]
I tried on latest and I get this exception when loading s3 files now:

```
Traceback (most recent call last):
  File "/home/jbe/.conda/envs/activeviam/lib/python3.7/site-packages/atoti/exceptions.py", line 219, in catch_py4j_exceptions
    return method(*args, **kwargs)
  File "/home/jbe/.conda/envs/activeviam/lib/python3.7/site-packages/atoti/_java_api.py", line 366, in create_store_from_csv
    store_name, store_params, load_params, csv_params
  File "/home/jbe/.conda/envs/activeviam/lib/python3.7/site-packages/atoti/vendor/atotipy4j/java_gateway.py", line 1314, in __call__
    answer, self.gateway_client, self.target_id, self.name)
  File "/home/jbe/.conda/envs/activeviam/lib/python3.7/site-packages/atoti/vendor/atotipy4j/protocol.py", line 324, in get_return_value
    format(target_id, ".", name), value)
Py4JJavaError: An error occurred while calling o2.createStoreFromCsv.
: com.amazonaws.SdkClientException: Unable to load AWS credentials from any provider in the chain: [EnvironmentVariableCredentialsProvider: Unable to load AWS credentials from environment variables (AWS_ACCESS_KEY_ID (or AWS_ACCESS_KEY) and AWS_SECRET_KEY (or AWS_SECRET_ACCESS_KEY)), SystemPropertiesCredentialsProvider: Unable to load AWS credentials from Java system properties (aws.accessKeyId and aws.secretKey), com.amazonaws.auth.profile.ProfileCredentialsProvider@fe73998: profile file cannot be null, com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper@940781b: Failed to connect to service endpoint: ]
	at com.amazonaws.auth.AWSCredentialsProviderChain.getCredentials(AWSCredentialsProviderChain.java:136)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.getCredentialsFromContext(AmazonHttpClient.java:1239)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.runBeforeRequestHandlers(AmazonHttpClient.java:815)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:765)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:752)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:726)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:686)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:668)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:532)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:512)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5052)
	at com.amazonaws.services.s3.AmazonS3Client.getBucketRegionViaHeadRequest(AmazonS3Client.java:5840)
	at com.amazonaws.services.s3.AmazonS3Client.fetchRegionFromCache(AmazonS3Client.java:5813)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5036)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4998)
	at com.amazonaws.services.s3.AmazonS3Client.getObjectMetadata(AmazonS3Client.java:1335)
	at com.amazonaws.services.s3.AmazonS3Client.getObjectMetadata(AmazonS3Client.java:1309)
	at com.amazonaws.services.s3.AmazonS3Client.doesObjectExist(AmazonS3Client.java:1390)
	at com.activeviam.chouket.util.files.S3Path.exist(S3Path.java:160)
	at com.activeviam.chouket.loading.impl.AFileBasedDataTable.<init>(AFileBasedDataTable.java:44)
	at com.activeviam.chouket.loading.csv.impl.CsvDataTable.<init>(CsvDataTable.java:73)
	at com.activeviam.chouket.loading.csv.impl.SimpleCsvDataSource.<init>(SimpleCsvDataSource.java:46)
	at com.activeviam.chouket.api.impl.ActiveViamSessionApiImpl.createStoreFromCsv(ActiveViamSessionApiImpl.java:132)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:834)


During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "<ipython-input-3-c0cabd5af8cc>", line 4, in <module>
    store_name="asset_positions_store",
  File "/home/jbe/.conda/envs/activeviam/lib/python3.7/site-packages/atoti/_docs_utils.py", line 18, in wrapper
    return func(*args, **kwargs)
  File "/home/jbe/.conda/envs/activeviam/lib/python3.7/site-packages/atoti/session.py", line 474, in read_csv
    mode,
  File "/home/jbe/.conda/envs/activeviam/lib/python3.7/site-packages/atoti/exceptions.py", line 222, in catch_py4j_exceptions
    "An error occurred in the JVM:", str(java_exception), java_exception
AtotiJavaException: An error occurred in the JVM:
An error occurred while calling o2.createStoreFromCsv.
: com.amazonaws.SdkClientException: Unable to load AWS credentials from any provider in the chain: [EnvironmentVariableCredentialsProvider: Unable to load AWS credentials from environment variables (AWS_ACCESS_KEY_ID (or AWS_ACCESS_KEY) and AWS_SECRET_KEY (or AWS_SECRET_ACCESS_KEY)), SystemPropertiesCredentialsProvider: Unable to load AWS credentials from Java system properties (aws.accessKeyId and aws.secretKey), com.amazonaws.auth.profile.ProfileCredentialsProvider@fe73998: profile file cannot be null, com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper@940781b: Failed to connect to service endpoint: ]
	at com.amazonaws.auth.AWSCredentialsProviderChain.getCredentials(AWSCredentialsProviderChain.java:136)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.getCredentialsFromContext(AmazonHttpClient.java:1239)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.runBeforeRequestHandlers(AmazonHttpClient.java:815)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:765)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:752)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:726)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:686)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:668)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:532)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:512)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5052)
	at com.amazonaws.services.s3.AmazonS3Client.getBucketRegionViaHeadRequest(AmazonS3Client.java:5840)
	at com.amazonaws.services.s3.AmazonS3Client.fetchRegionFromCache(AmazonS3Client.java:5813)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5036)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4998)
	at com.amazonaws.services.s3.AmazonS3Client.getObjectMetadata(AmazonS3Client.java:1335)
	at com.amazonaws.services.s3.AmazonS3Client.getObjectMetadata(AmazonS3Client.java:1309)
	at com.amazonaws.services.s3.AmazonS3Client.doesObjectExist(AmazonS3Client.java:1390)
	at com.activeviam.chouket.util.files.S3Path.exist(S3Path.java:160)
	at com.activeviam.chouket.loading.impl.AFileBasedDataTable.<init>(AFileBasedDataTable.java:44)
	at com.activeviam.chouket.loading.csv.impl.CsvDataTable.<init>(CsvDataTable.java:73)
	at com.activeviam.chouket.loading.csv.impl.SimpleCsvDataSource.<init>(SimpleCsvDataSource.java:46)
	at com.activeviam.chouket.api.impl.ActiveViamSessionApiImpl.createStoreFromCsv(ActiveViamSessionApiImpl.java:132)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by fabiencelier** [comment]
This is fixed in 0.4.0, thanks for reporting and testing
